Packet++/header/GvcpLayer.h

```diff
@@ -0,0 +1,626 @@
+	{
+		// Discovery Protocol Control
+		DiscoveredCmd = 0x0002,
+		DiscoveredAck = 0x0003,
+		ForceIpCmd = 0x0004,
+		ForceIpAck = 0x0005,
+
+		// Streaming Protocol Control
+		PacketResendCmd = 0x0040,
+		PacketResendAck = 0x0041,  // Resent packet must be on the stream channel
+
+		// Device Memory Access
+		ReadRegCmd = 0x0080,
+		ReadRegAck = 0x0081,
+		WriteRegCmd = 0x0082,
+		WriteRegAck = 0x0083,
+		ReadMemCmd = 0x0084,
+		ReadMemAck = 0x0085,
+		WriteMemCmd = 0x0086,
+		WriteMemAck = 0x0087,
+		PendingAck = 0x0089,
+
+		// Asynchronous Events
+		EventCmd = 0x00C0,
+		EventAck = 0x00C1,
+		EventDataCmd = 0x00C2,
+		EventDataAck = 0x00C3,
+
+		// Miscellaneous
+		ActionCmd = 0x0100,
+		ActionAck = 0x0101,
+		Unknown = 0xFFFF
+	};
+
+	std::ostream& operator<<(std::ostream& os, GvcpCommand command);
+
+	/// @brief Gvcp response status
+	/// See spec "Table 19-1: List of Standard Status Codes"
+	enum class GvcpResponseStatus : uint16_t
+	{
+		Success = 0x0000,
+		PacketResend = 0x0100,
+		NotImplemented = 0x8001,
+		InvalidParameter = 0x8002,
+		InvalidAddress = 0x8003,
+		WriteProtect = 0x8004,
+		BadAlignment = 0x8005,
+		AccessDenied = 0x8006,
+		Busy = 0x8007,
+		LocalProblem = 0x8008,     // deprecated
+		MsgMismatch = 0x8009,      // deprecated
+		InvalidProtocol = 0x800A,  // deprecated
+		NoMsg = 0x800B,            // deprecated
+		PacketUnavailable = 0x800C,
+		DataOverrun = 0x800D,
+		InvalidHeader = 0x800E,
+		WrongConfig = 0x800F,  // deprecated
+		PacketNotYetAvailable = 0x8010,
+		PacketAndPrevRemovedFromMemory = 0x8011,
+		PacketRemovedFromMemory = 0x8012,
+		NoRefTime = 0x8013,                     // GEV 2.0
+		PacketTemporarilyUnavailable = 0x8014,  // GEV 2.0
+		Overflow = 0x8015,                      // GEV 2.0
+		ActionLate = 0x8016,                    // GEV 2.0
+		LeaderTrailerOverflow = 0x8017,         // GEV 2.1
+		Error = 0x8FFF,
+		Unknown = 0xFFFF
+	};
+
+	std::ostream& operator<<(std::ostream& os, GvcpResponseStatus status);
+
+#pragma pack(push, 1)
+	/// @brief Gvcp request header
+	/// @note refer to the spec "15.1 Request Header". The data is stored as big-endian.
+	struct GvcpRequestHeader
+	{
+		friend class GvcpRequestLayer;
+
+	protected:
+		uint8_t magicNumber = internal::kGvcpMagicNumber;  // always fixed
+		uint8_t flag = 0;  // 0-3 bits are specified by each command, 4-6 bits are reserved, 7 bit is acknowledge
+		uint16_t command = 0;
+		uint16_t dataSize = 0;
+		uint16_t requestId = 0;
+
+	public:
+		// ------------- methods --------------
+		GvcpRequestHeader() = default;
+
+		GvcpRequestHeader(GvcpFlag flag, GvcpCommand command, uint16_t dataSize, uint16_t requestId)
+		    : flag(flag), command(hostToNet16(static_cast<uint16_t>(command))), dataSize(hostToNet16(dataSize)),
+		      requestId(hostToNet16(requestId))
+		{}
+
+		GvcpFlag getFlag() const
+		{
+			return flag;
+		}
+
+		GvcpCommand getCommand() const
+		{
+			return static_cast<GvcpCommand>(netToHost16(command));
```

Comment:
I think we can move all implementations that use `netToHost16` to `GvcpLayer.cpp` so we don't expose `SystemUtils.h` in the header file

Reply:
Since it's one line function, I think it's clearer to be in the header. I forward declare the function, so we don't need to include the whole header file.